text/0165-instrumentation-api.md

```diff
@@ -0,0 +1,236 @@
+spans or recording metrics. However, this is only half the story when it comes to instrumenting and
+misses out on common patterns required during instrumentation of libraries. Without defining these
+patterns through an instrumentation API, it is difficult to ensure consistency of quality and user
+experience of different instrumentation libraries.
+
+The goals of the instrumentation API are
+
+- Provide a good user experience for authors of library instrumentation that minimizes logic and cognitive
+load.
+
+- Codify quality requirements of good library instrumentation, for example completeness of semantic
+conventions.
+
+- Provide a consistent user experience for users of library instrumentation to ensure similar
+configurability and usability regardless of the library being used.
+
+## Explanation
+
+The instrumentation API is a higher level API built on top of the OpenTelemetry API. The key notion
+is that instrumentation of libraries almost always involves intercepting a library at two points, the
+start and end of an operation, for example an HTTP request or a DB query. This means that the primary
+instrumentation API only needs to provide two methods, `start` and `end` which are called from
+the instrumentation at the beginning and end of the operation. This API is called the `Instrumenter`.
+
+In addition to the operation lifecycle, the other library-specific aspect of instrumentation is
+mapping a library domain object into OpenTelemetry semantic conventions. The instrumentation API
+allows instrumentation to implement `Extractor`s, which map from a request or response object into
+attributes. Extractors enforce a contract, for example by defining an abstract class, which includes
+all the semantic conventions of a given namespace in OpenTelemetry. The library instrumentation just
+implements extractors relevant to the library - for example, an HTTP framework will implement the
+HTTP extractor. This allows all HTTP attributes to be filled without specific knowledge of the
+semantic conventions themselves.
+
+An `InstrumentaterBuilder` stitches together `Instrumenter` and `Extractor`s, as well as consistently
+defining a set of well known configuration knobs for library users. The instrumentation library will
+initialize a builder with defaults registering the implemented `Extractor`s and return it to the
+user of the instrumentation, which can further configure it to their needs.
+
+## Internal details
+
+These details are based on a [prototype](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/9f31a057b6878619157aff93e27fdb15fad89958/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter)
+in Java.
+
+### Instrumenter
+
+The `Instrumenter` is the main entry point for the author of library instrumentation. It has two
+primary methods.
+
+- `Context start(Context parentContext, REQUEST request)`: The method to invoke at the beginning of an
```

Comment:
The instrumentation API is meant to make standard telemetry instrumentation easier, I could try to write that but a bit unsure the best way to phrase it. Internal implementation details like a view may be best with the OpenTelemetry API especially if it doesn't semantic conventions - the main reason to use a mapping of request to attributes is to enforce the conventions. Internal instrumentation also doesn't need to worry about propagation, nesting, etc that this API tries to tackle

Reply:
I mentioned that specific example because I've been working on optional auto-instrumentation for the `opentelemetry-ruby` project that instruments parts of Rails like that automatically, and I think other frameworks might fall into that category too. I agree that end-users need not use this Instrumentation API, but people like me (writing auto-instrumentation for others) would. That's the kind of thing I had in mind, if that helps clarify why I mentioned it?